uarray/backend.py

```diff
@@ -9,11 +9,62 @@
 ArgumentReplacerType = Callable[[Tuple, Dict, Tuple], Tuple[Tuple, Dict]]
 
 
-class BackendNotImplementedError(TypeError):
-    pass
+class BackendNotImplementedError(NotImplementedError):
+    """
+    An exception that is thrown when no compatible backend is found for a method.
+    """
 
 
 class MultiMethod:
+    """
+    This class is meant to provide a dispatchable method with or without a registered implementation.
+
+    Provides a multi-method instance. The method has no implementation by default, but one can be provided
+    if needed. :obj:`Backend` s can provide implementations for the method, and users can choose backends.
+
+    Parameters
+    ----------
+    argument_extractor : ArgumentExtractorType
+        This is a callable that extracts the arguments using which dispatch can be performed from this
+        multimethod.
+    argument_replacer : ArgumentReplacerType
+        This takes in args, kwargs and dispatchable args, and replaces all dispatchable arguments within
+        the args and kwargs, and then returns them.
+    default : Optional[Callable]
```

Comment:
minor: the standard way to write this is `callable, optional`.

Reply:
Hmm, I'm confused now. Further down there's `Optional[bool], optional`. If `Optional` is a class then I don't see it so quickly.

Reply:
@rgommers `typing.Optional[T]` basically means that you can have `None` in addition to the allowed values of `T`. The `, optional` means that it's an optional argument in itself.